README.md
<!--
SPDX-FileCopyrightText: 2016-2021 Comcast Cable Communications Management, LLC
SPDX-License-Identifier: Apache-2.0
-->
# trower-base64

C implementation of base64 encode/decode.

[![Build Status](https://github.com/xmidt-org/trower-base64/workflows/CI/badge.svg)](https://github.com/xmidt-org/trower-base64/actions)
[![codecov.io](http://codecov.io/github/xmidt-org/trower-base64/coverage.svg?branch=master)](http://codecov.io/github/xmidt-org/trower-base64?branch=master)
[![coverity](https://img.shields.io/coverity/scan/23253.svg)](https://scan.coverity.com/projects/xmidt-org-trower-base64)
[![Quality Gate Status](https://sonarcloud.io/api/project_badges/measure?project=xmidt-org_trower-base64&metric=alert_status)](https://sonarcloud.io/dashboard?id=xmidt-org_trower-base64)
[![Language Grade: C/C++](https://img.shields.io/lgtm/grade/cpp/g/xmidt-org/trower-base64.svg?logo=lgtm&logoWidth=18)](https://lgtm.com/projects/g/xmidt-org/trower-base64/context:cpp)
[![GitHub release](https://img.shields.io/github/release/xmidt-org/trower-base64.svg)](CHANGELOG.md)

This started as a fork of the implementation Bob Trower produced and released in
2001.  You can find the original here: http://base64.sourceforge.net/

## Why fork?

As we started using the library we found it was hugely handy, but had a few bugs
and not as many tests as we wanted.  As the upstream project seems to have lost
steam, we decided to fork, improve, release and ship.

# Building and Testing Instructions

```
meson build
cd build
ninja all test coverage
firefox meson-logs/coveragereport/index.html
```


CHANGELOG.md
<!--
SPDX-FileCopyrightText: 2016-2021 Comcast Cable Communications Management, LLC
SPDX-License-Identifier: Apache-2.0
-->
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [v1.2.7]
- Add a meson wrap file to the release objects.
- Improve the meson build so normally tests are run with valgrind but if this is
  a subproject then valgrind and the tests are disabled.

## [v1.2.3]
- Fix the include path so it works automatically.

## [v1.2.2]
- Improve include structure for meson.

## [v1.2.1]
- Move to use meson for building.
- Remove cmake.

## [v1.2.0]
- Add a few helper functions to allocate buffers and fill them.
- Add a `b64_encode()` function.

## [v1.1.5]
- Rewrite the core handling to fix a number of defects found in 1.1.4.
- Improve the tests to better prove the library completely works.

## [v1.1.4]
- Update the license to use SPDX
- Update the build infrastructure

## [v1.1.3]
- Added a more durable release artifact process.
- Fixed a few code smells according to sonarcloud.io

## [v1.1.2] - 2019-01-07
- Added valgrind to travis-ci linux builds
- Fixed output buffer overrun error due to incorrect buffer size answer
- Fixed output buffer overrun error due to limited characters
- Fixed input buffer read overrun error

## [v1.1.1] - 2018-10-18
- Added travis-ci
- Travis build for MacOSX
- Fixed build on Mac and Centos
- Added APIs for base64url buffer sizes
- Uses CTest module to build with test disabled
- Add the trower-base64 directory into the include to fix consumers of this project
- Changed cmake minimum version required from 3.3.2 to 2.8.7

## [v1.1] - 2016-02-15
### Added
- Add base64url decode implementaion
### Changed
- Improve cmake files

## [v1.0] - 2016-02-15
### Added
- Initial creation

## [0.1] - 2016-02-05
### Added
- Initial creation

[Unreleased]: https://github.com/xmidt-org/trower-base64/compare/v1.2.7...HEAD
[v1.2.7]: https://github.com/xmidt-org/trower-base64/compare/v1.2.3...v1.2.7
[v1.2.3]: https://github.com/xmidt-org/trower-base64/compare/v1.2.2...v1.2.3
[v1.2.2]: https://github.com/xmidt-org/trower-base64/compare/v1.2.1...v1.2.2
[v1.2.1]: https://github.com/xmidt-org/trower-base64/compare/v1.2.0...v1.2.1
[v1.2.0]: https://github.com/xmidt-org/trower-base64/compare/v1.1.5...v1.2.0
[v1.1.5]: https://github.com/xmidt-org/trower-base64/compare/v1.1.4...v1.1.5
[v1.1.4]: https://github.com/xmidt-org/trower-base64/compare/v1.1.3...v1.1.4
[v1.1.3]: https://github.com/xmidt-org/trower-base64/compare/v1.1.2...v1.1.3
[v1.1.2]: https://github.com/xmidt-org/trower-base64/compare/v1.1.1...v1.1.2
[v1.1.1]: https://github.com/xmidt-org/trower-base64/compare/v1.1...v1.1.1
[v1.1]: https://github.com/xmidt-org/trower-base64/compare/v1.0...v1.1
[v1.0]: https://github.com/xmidt-org/trower-base64/compare/3e997140737790d284de03a5ef6f497798673750...v1.0


src/b64.c
/* SPDX-FileCopyrightText: 2010-2021 Comcast Cable Communications Management, LLC */
/* SPDX-License-Identifier: Apache-2.0 */
/* SPDX-FileCopyrightText: 2001 Bob Trower, Trantor Standard Systems Inc. */
/* SPDX-License-Identifier: MIT */

/*********************************************************************\

MODULE NAME:    b64.c

AUTHOR:         Bob Trower 08/04/01

PROJECT:        Crypt Data Packaging

DESCRIPTION:
                This little utility implements the Base64
                Content-Transfer-Encoding standard described in
                RFC1113 (http://www.faqs.org/rfcs/rfc1113.html).

                This is the coding scheme used by MIME to allow
                binary data to be transferred by SMTP mail.

                Groups of 3 bytes from a binary stream are coded as
                groups of 4 bytes in a text stream.

                The input stream is 'padded' with zeros to create
                an input that is an even multiple of 3.

                A special character ('=') is used to denote padding so
                that the stream can be decoded back to its exact size.

                Encoded output is formatted in lines which should
                be a maximum of 72 characters to conform to the
                specification.  This program defaults to 72 characters,
                but will allow more or less through the use of a
                switch.  The program enforces a minimum line size
                of 4 characters.

                Example encoding:

                The stream 'ABCD' is 32 bits long.  It is mapped as
                follows:

                ABCD

                 A (65)     B (66)     C (67)     D (68)   (None) (None)
                01000001   01000010   01000011   01000100

                16 (Q)  20 (U)  9 (J)   3 (D)    17 (R) 0 (A)  NA (=) NA (=)
                010000  010100  001001  000011   010001 000000 000000 000000


                QUJDRA==

                Decoding is the process in reverse.  A 'decode' lookup
                table has been created to avoid string scans.

DESIGN GOALS:   Specifically:
        Code is a stand-alone utility to perform base64
        encoding/decoding. It should be genuinely useful
        when the need arises and it meets a need that is
        likely to occur for some users.
        Code acts as sample code to show the author's
        design and coding style.

        Generally:
        This program is designed to survive:
        Everything you need is in a single source file.
        It compiles cleanly using a vanilla ANSI C compiler.
        It does its job correctly with a minimum of fuss.
        The code is not overly clever, not overly simplistic
        and not overly verbose.
        Access is 'cut and paste' from a web page.
        Terms of use are reasonable.

VALIDATION:     Non-trivial code is never without errors.  This
                file likely has some problems, since it has only
                been tested by the author.  It is expected with most
                source code that there is a period of 'burn-in' when
                problems are identified and corrected.  That being
                said, it is possible to have 'reasonably correct'
                code by following a regime of unit test that covers
                the most likely cases and regression testing prior
                to release.  This has been done with this code and
                it has a good probability of performing as expected.

                Unit Test Cases:

                case 0:empty file:
                    CASE0.DAT  ->  ->
                    (Zero length target file created
                    on both encode and decode.)

                case 1:One input character:
                    CASE1.DAT A -> QQ== -> A

                case 2:Two input characters:
                    CASE2.DAT AB -> QUJD -> AB

                case 3:Three input characters:
                    CASE3.DAT ABC -> QUJD -> ABC

                case 4:Four input characters:
                    case4.dat ABCD -> QUJDRA== -> ABCD

                case 5:All chars from 0 to ff, linesize set to 50:

                    AAECAwQFBgcICQoLDA0ODxAREhMUFRYXGBkaGxwdHh8gISIj
                    JCUmJygpKissLS4vMDEyMzQ1Njc4OTo7PD0+P0BBQkNERUZH
                    SElKS0xNTk9QUVJTVFVWV1hZWltcXV5fYGFiY2RlZmdoaWpr
                    bG1ub3BxcnN0dXZ3eHl6e3x9fn+AgYKDhIWGh4iJiouMjY6P
                    kJGSk5SVlpeYmZqbnJ2en6ChoqOkpaanqKmqq6ytrq+wsbKz
                    tLW2t7i5uru8vb6/wMHCw8TFxsfIycrLzM3Oz9DR0tPU1dbX
                    2Nna29zd3t/g4eLj5OXm5+jp6uvs7e7v8PHy8/T19vf4+fr7
                    /P3+/w==

                case 6:Mime Block from e-mail:
                    (Data same as test case 5)

                case 7: Large files:
                    Tested 28 MB file in/out.

                case 8: Random Binary Integrity:
                    This binary program (b64.exe) was encoded to base64,
                    back to binary and then executed.

                case 9 Stress:
                    All files in a working directory encoded/decoded
                    and compared with file comparison utility to
                    ensure that multiple runs do not cause problems
                    such as exhausting file handles, tmp storage, etc.

                -------------

                Syntax, operation and failure:
                    All options/switches tested.  Performs as
                    expected.

                case 10:
                    No Args -- Shows Usage Screen
                    Return Code 1 (Invalid Syntax)
                case 11:
                    One Arg (invalid) -- Shows Usage Screen
                    Return Code 1 (Invalid Syntax)
                case 12:
                    One Arg Help (-?) -- Shows detailed Usage Screen.
                    Return Code 0 (Success -- help request is valid).
                case 13:
                    One Arg Help (-h) -- Shows detailed Usage Screen.
                    Return Code 0 (Success -- help request is valid).
                case 14:
                    One Arg (valid) -- Uses stdin/stdout (filter)
                    Return Code 0 (Sucess)
                case 15:
                    Two Args (invalid file) -- shows system error.
                    Return Code 2 (File Error)
                case 16:
                    Encode non-existent file -- shows system error.
                    Return Code 2 (File Error)
                case 17:
                    Out of disk space -- shows system error.
                    Return Code 3 (File I/O Error)

                -------------

                Compile/Regression test:
                    gcc compiled binary under Cygwin
                    Microsoft Visual Studio under Windows 2000
                    Microsoft Version 6.0 C under Windows 2000

DEPENDENCIES:   None

VERSION HISTORY:
                Bob Trower 08/04/01 -- Create Version 0.00.00B
                Trent Schmidt 2/7/2010 -- Modified code heavily for our use
                Craig Barnes 4/2/2013 -- Completely reworked for space savings.

\******************************************************************* */

#include "base64.h"

/**
 *  Get the size of the buffer required to hold the decoded data when base64 url encoded.
 *  Not supported
 */
size_t b64url_get_encoded_buffer_size( const size_t decoded_size )
{
	return decoded_size;
}

/**
 *  Get the size of the buffer required to hold the decoded data when base64 encoded.
 */
size_t b64_get_encoded_buffer_size( const size_t decoded_size )
{
    return ((decoded_size + 2) / 3) << 2;
}


/*
** encode
**
** base64-encode a stream adding padding and line breaks as per spec.
 * Assume that output is large enough to fill out last 4-byte boundary.
*/
void b64_encode( const uint8_t *input, const size_t input_size, uint8_t *output )
{
    size_t remsize = input_size;    // (Note: could just use input_size were it not const)
    const uint8_t *last = &input[input_size];
    while (remsize > 0) {
        int i;
        uint32_t tmp;
        int len = (remsize >= 3) ? 3 : (int) remsize; // remsize is [0-4]
        remsize -= len;
        tmp = 0;
        for (i=0; i < 3; ++i) {
            tmp <<= 8;
            // Don't read past the end of the input buffer
            if (input < last) {
                tmp |= *input++;
            }
        }
        for (i=3; i >= 0; i--) {
            // Translation Table as described in RFC1113
            const char map[]= "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/";

            uint8_t index = (uint8_t) (tmp & 0x3F);
            // Note: per earlier code, we only use '=' in at most the last 2 output bytes
            //  The following comparison should still do this (len is 1-3)
            if (i > len) {
                output[i] = '=';   // use '=' for chars past end of input
            } else {
                output[i] = (uint8_t) map[index];
            }
            tmp >>= 6;
        }
        output += 4;
    }
}


src/base64.h
/* SPDX-FileCopyrightText: 2010-2021 Comcast Cable Communications Management, LLC */
/* SPDX-License-Identifier: Apache-2.0 */
#ifndef __BASE64_LIB__
#define __BASE64_LIB__

#ifdef __cplusplus
extern "C" {
#endif

#include <stdint.h>
#include <stddef.h>

/*----------------------------------------------------------------------------*/
/*                             Standard Base64                                */
/*----------------------------------------------------------------------------*/

/**
 *  Get the size of the buffer required to hold the decoded data when encoded.
 *
 *  @note: The size returned does not account for any trailing '\0'.
 *
 *  @param decoded_size size of the decoded date
 *
 *  @return size of the buffer required to hold the encoded data
 */
size_t b64_get_encoded_buffer_size( const size_t decoded_size );


/**
 * Get the size of the buffer needed to hold the decoded output.
 *
 * @note: The size MAY be larger the the resulting decoded output.
 *
 * @param encoded_size size of the encoded data
 *
 * @return size of the raw data
 */
size_t b64_get_decoded_buffer_size( const size_t encoded_size );


/**
 *  Encodes the input into base64.  The base 64 produced string will be placed
 *  into the output param.  Consumers of this function are responsible for
 *  making sure the output buffer is able to hold all of the data once it is
 *  encoded.
 *
 *  @note: The output buffer must be large enough to handle the encoded payload.
 *
 *  @param raw  pointer to the raw data
 *  @param len  size of the raw data in bytes
 *  @param out  pointer to where the encoded data should be placed
 */
void b64_encode( const uint8_t *raw, const size_t len, uint8_t *out );


/**
 * Decodes the base64 encoded buffer.  The produced raw buffer will be placed
 * into the output array.  Consumers of this function are responsible for making
 * sure the output buffer is large enough to hold all of the decoded data.
 *
 * @note: The output buffer must be large enough to handle the decoded payload.
 *
 * @param enc  pointer to the encoded data
 * @param len  size of the raw data
 * @param out  pointer to where the decoded data should be placed
 *
 * @return total number of bytes in the decoded array, or 0 if there was a
 *         decoding error
 */
size_t b64_decode( const uint8_t *enc, const size_t len, uint8_t *out );


/**
 * Decodes the base64 buffer into a new buffer with the size specified in
 * out_len.
 *
 * @note: The returned buffer must have free() called to prevent a memory
 *        leak.
 *
 * @param enc      pointer to the encoded data
 * @param len      size of the raw data
 * @param out_len  pointer to where the resulting buffer length is placed
 *
 * @return the buffer containing the raw bytes or NULL on error
 */
uint8_t* b64_decode_with_alloc( const uint8_t *enc, size_t len, size_t *out_len );


/**
 * Encodes the raw bytes using standard base64 into a new buffer with the size
 * specified in out_len.
 *
 * @note: The returned buffer must have free() called to prevent a memory
 *        leak.
 * @note: The returned buffer is '\0' terminated.
 * @note: The out_len value is equivalent to strlen() of the returned buffer.
 *
 * @param raw      pointer to the encoded data
 * @param len      size of the raw data
 * @param out_len  pointer to where the resulting buffer length is placed
 *
 * @return the buffer containing the raw bytes or NULL on error
 */
char* b64_encode_with_alloc( const uint8_t *raw, size_t len, size_t *out_len );


/*----------------------------------------------------------------------------*/
/*                                 URL Base64                                 */
/*----------------------------------------------------------------------------*/


/**
 *
 *  Get the size of the buffer required to hold the decoded data when base64
 *  url encoded.
 *
 *  @note: The size returned does not account for any trailing '\0'.
 *
 *  @param decoded_size size of the decoded date
 *
 *  @return size of the buffer required to hold the encoded data
 */
size_t b64url_get_encoded_buffer_size( const size_t decoded_size );


/**
 * Get the size of the buffer needed to hold the output decoded from
 * base64url encoded data.
 *
 * @note: The size MAY be larger the the resulting decoded output.
 *
 * @param encoded_size size of the base64url encoded data
 *
 * @return size of the raw data
 */
size_t b64url_get_decoded_buffer_size( const size_t encoded_size );


/**
 *  Encodes the input into base64url.  The base 64 produced string will be placed
 *  into the output param.  Consumers of this function are responsible for
 *  making sure the output buffer is able to hold all of the data once it is
 *  encoded.
 *
 *  @note: The output buffer must be large enough to handle the encoded payload.
 *
 *  @param raw pointer to the raw data
 *  @param len size of the raw data in bytes
 *  @param output pointer to where the encoded data should be placed
 */
void b64url_encode( const uint8_t *raw, const size_t len, uint8_t *output );


/**
 * Decodes the base64url encoded stream.  The produced raw buffer will be
 * placed into the output array.  Consumers of this function are responsible for
 * making sure the output buffer is large enough to hold all of the decoded data.
 *
 * @note: The output buffer must be large enough to handle the decoded payload.
 *
 * @param enc pointer to the encoded data
 * @param len size of the raw data
 * @param output pointer to where the decoded data should be placed
 *
 * @return total number of bytes in the decoded array, or 0 if there was a
 *         decoding error
 */
size_t b64url_decode( const uint8_t *enc, const size_t len, uint8_t *output );


/**
 * Decodes the base64url buffer into a new buffer with the size specified in
 * out_len.
 *
 * @note: The returned buffer must have free() called to prevent a memory
 *        leak.
 *
 * @param enc      pointer to the encoded data
 * @param len      size of the raw data
 * @param out_len  pointer to where the resulting buffer length is placed
 *
 * @return the buffer containing the raw bytes or NULL on error
 */
uint8_t* b64url_decode_with_alloc( const uint8_t *enc, size_t len, size_t *out_len );


/**
 * Encodes the raw bytes using base64url into a new buffer with the size
 * specified in out_len.
 *
 * @note: The returned buffer must have free() called to prevent a memory
 *        leak.
 * @note: The returned buffer is '\0' terminated.
 * @note: The out_len value is equivalent to strlen() of the returned buffer.
 *
 * @param raw      pointer to the encoded data
 * @param len      size of the raw data
 * @param out_len  pointer to where the resulting buffer length is placed
 *
 * @return the buffer containing the raw bytes or NULL on error
 */
char* b64url_encode_with_alloc( const uint8_t *raw, size_t len, size_t *out_len );


#ifdef __cplusplus
}
#endif

#endif /* __BASE64_LIB__ */


.github/scripts/make_release_artifacts.sh
#!/bin/bash
# SPDX-FileCopyrightText: 2021 Comcast Cable Communications Management, LLC
# SPDX-License-Identifier: Apache-2.0

function how_to_use {
    echo "Usage: $0 -v version -r repo [-m] [-p provides]"
    echo "	-v version      Example: (v1.2.3)"
    echo "	-r github repo  Example: (octocat/Hello-World)"
    echo "	-m if passed in generates a meson wrap file"
    echo "	-p the meson library provided or the project name by default Example: foo -> libfoo = libfoo_dep"
}

version=none
repo_slug=none
provides=none
meson=0

while getopts ":v:r:pmh:" opt; do
    case ${opt} in
        v ) version=${OPTARG} ;;
        r ) repo_slug=${OPTARG} ;;
        p ) provides=${OPTARG} ;;
        m ) meson=1 ;;
        h )
            how_to_use
            exit 0
            ;;
        \?) how_to_use
            exit 0
            ;;
        : ) how_to_use
            exit 1
            ;;
    esac
done

if [ "none" == $version -o "none" == $repo_slug ]; then
    how_to_use
    exit 1
fi

raw_version=$version
if [ "v" == ${version:0:1} ]; then
    raw_version=${version:1}
fi

repo_name=$(echo $repo_slug | awk -F '/' '{print $2}')
release_slug=$(echo "$repo_name-$raw_version")

if [ "none" == $provides ]; then
    provides=$repo_name
fi

echo "     version: $version"
echo " raw_version: $raw_version"
echo "   repo_slug: $repo_slug"
echo "   repo_name: $repo_name"
echo "release_slug: $release_slug"
echo "    provides: $provides"

echo ""
echo "Making the .tar.gz archive"
git archive --format=tar.gz -o $release_slug.tar.gz --prefix=$release_slug/ $version

echo "Making the .zip archive"
git archive --format=zip    -o $release_slug.zip --prefix=$release_slug/ $version

if [ 1 == $meson ]; then

    echo "Making the .wrap file for meson"

    tgz_sum=`sha256sum $release_slug.tar.gz`
    tgz_sum=${tgz_sum:0:64} # Keep only the first 64 bytes which are the checksum

    echo "[wrap-file]"                               > $repo_name.wrap
    echo "directory = $release_slug"                >> $repo_name.wrap
    echo ""                                         >> $repo_name.wrap
    echo "source_filename = $release_slug.tar.gz"   >> $repo_name.wrap
    echo "source_url = https://github.com/$repo_slug/releases/download/$version/$release_slug.tar.gz" >> $repo_name.wrap
    echo "source_hash = $tgz_sum"                   >> $repo_name.wrap
    echo ""                                         >> $repo_name.wrap
    echo "[provides]"                               >> $repo_name.wrap
    echo "lib$provides = lib${provides}_dep"        >> $repo_name.wrap

    echo "Making the sha256sums.txt file"
    sha256sum $release_slug.tar.gz $release_slug.zip $repo_name.wrap > $release_slug-sha256sums.txt
else
    echo "Making the sha256sums.txt file"
    sha256sum $release_slug.tar.gz $release_slug.zip > $release_slug-sha256sums.txt
fi

echo "Copying files to the artifacts directory"
mkdir artifacts
cp ${release_slug}* artifacts/.

if [ -f $repo_name.wrap ]; then
    cp $repo_name.wrap artifacts/.
fi

echo "Complete"


src/base64.c
/* SPDX-FileCopyrightText: 2021 Comcast Cable Communications Management, LLC */
/* SPDX-License-Identifier: Apache-2.0 */

#include <stddef.h>
#include <stdint.h>
#include <stdlib.h>

/*----------------------------------------------------------------------------*/
/*                                   Macros                                   */
/*----------------------------------------------------------------------------*/
/* none */

/*----------------------------------------------------------------------------*/
/*                               Data Structures                              */
/*----------------------------------------------------------------------------*/
/* none */

/*----------------------------------------------------------------------------*/
/*                            File Scoped Variables                           */
/*----------------------------------------------------------------------------*/
/* none */

/*----------------------------------------------------------------------------*/
/*                             Function Prototypes                            */
/*----------------------------------------------------------------------------*/
static void encode( const char *map, const uint8_t *in, size_t len, uint8_t *out );
static size_t decode( const int8_t *map, const uint8_t *in, size_t len, uint8_t *out );
static uint8_t* decode_w_alloc( size_t (size_fn)(const size_t),
                                size_t (decode_fn)(const uint8_t*, const size_t, uint8_t*),
                                const uint8_t *enc, size_t len, size_t *out_len );
static char* encode_w_alloc( size_t (size_fn)(const size_t),
                             void (encode_fn)(const uint8_t*, const size_t, uint8_t*),
                             const uint8_t *enc, size_t len, size_t *out_len );

/*----------------------------------------------------------------------------*/
/*                             External Functions                             */
/*----------------------------------------------------------------------------*/
size_t b64_get_encoded_buffer_size( const size_t decoded_size )
{
    return ((decoded_size + 2) / 3) * 4;
}


size_t b64_get_decoded_buffer_size( const size_t encoded_size )
{
    size_t rv = (encoded_size / 4) * 3;

    /* Check that the input is valid. */
    if( b64_get_encoded_buffer_size(rv) != encoded_size ) {
        rv = 0;
    }

    return rv;
}


size_t b64url_get_encoded_buffer_size( const size_t decoded_size )
{
    size_t remainder = decoded_size % 3;

    if( remainder ) {
        remainder++;
    }

    return (decoded_size / 3) * 4 + remainder;
}


size_t b64url_get_decoded_buffer_size( const size_t encoded_size )
{
    size_t remainder = 0x03 & encoded_size;
    size_t rv = 0;

    if( remainder ) {
        remainder--;
    }

    rv = (encoded_size / 4) * 3 + remainder;

    /* Check that the input is valid. */
    if( b64url_get_encoded_buffer_size(rv) != encoded_size ) {
        rv = 0;
    }

    return rv;
}


void b64_encode( const uint8_t *raw, const size_t len, uint8_t *out )
{
    static const char map[65] = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/=";

    encode( map, raw, len, out );
}


void b64url_encode( const uint8_t *raw, const size_t len, uint8_t *out )
{
    static const char map[65] = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789-_\0";

    encode( map, raw, len, out );
}


size_t b64_decode( const uint8_t *enc, const size_t len, uint8_t *out )
{
    // -1 = invalid
    // -2 = padding
    static const int8_t map[256] = {
        -1,-1,-1,-1, -1,-1,-1,-1, -1,-1,-1,-1, -1,-1,-1,-1,    /* 0x00-0x0f */
        -1,-1,-1,-1, -1,-1,-1,-1, -1,-1,-1,-1, -1,-1,-1,-1,    /* 0x10-0x1f */
        -1,-1,-1,-1, -1,-1,-1,-1, -1,-1,-1,62, -1,-1,-1,63,    /* 0x20-0x2f */
        52,53,54,55, 56,57,58,59, 60,61,-1,-1, -1,-2,-1,-1,    /* 0x30-0x3f */
        -1, 0, 1, 2,  3, 4, 5, 6,  7, 8, 9,10, 11,12,13,14,    /* 0x40-0x4f */
        15,16,17,18, 19,20,21,22, 23,24,25,-1, -1,-1,-1,-1,    /* 0x50-0x5f */
        -1,26,27,28, 29,30,31,32, 33,34,35,36, 37,38,39,40,    /* 0x60-0x6f */
        41,42,43,44, 45,46,47,48, 49,50,51,-1, -1,-1,-1,-1,    /* 0x70-0x7f */
        -1,-1,-1,-1, -1,-1,-1,-1, -1,-1,-1,-1, -1,-1,-1,-1,    /* 0x80-0x8f */
        -1,-1,-1,-1, -1,-1,-1,-1, -1,-1,-1,-1, -1,-1,-1,-1,    /* 0x90-0x9f */
        -1,-1,-1,-1, -1,-1,-1,-1, -1,-1,-1,-1, -1,-1,-1,-1,    /* 0xa0-0xaf */
        -1,-1,-1,-1, -1,-1,-1,-1, -1,-1,-1,-1, -1,-1,-1,-1,    /* 0xb0-0xbf */
        -1,-1,-1,-1, -1,-1,-1,-1, -1,-1,-1,-1, -1,-1,-1,-1,    /* 0xc0-0xcf */
        -1,-1,-1,-1, -1,-1,-1,-1, -1,-1,-1,-1, -1,-1,-1,-1,    /* 0xd0-0xdf */
        -1,-1,-1,-1, -1,-1,-1,-1, -1,-1,-1,-1, -1,-1,-1,-1,    /* 0xe0-0xef */
        -1,-1,-1,-1, -1,-1,-1,-1, -1,-1,-1,-1, -1,-1,-1,-1,    /* 0xf0-0xff */
    };
    size_t max = b64_get_decoded_buffer_size( len );

    if( (0 == max) || !enc || !out ) {
        return 0;
    }

    return decode( map, enc, len, out );
}


size_t b64url_decode( const uint8_t *enc, const size_t len, uint8_t *out )
{
    // -1 = invalid
    // -2 = padding
    static const int8_t map[256] = {
        -1,-1,-1,-1, -1,-1,-1,-1, -1,-1,-1,-1, -1,-1,-1,-1,    /* 0x00-0x0f */
        -1,-1,-1,-1, -1,-1,-1,-1, -1,-1,-1,-1, -1,-1,-1,-1,    /* 0x10-0x1f */
        -1,-1,-1,-1, -1,-1,-1,-1, -1,-1,-1,-1, -1,62,-1,-1,    /* 0x20-0x2f */
        52,53,54,55, 56,57,58,59, 60,61,-1,-1, -1,-2,-1,-1,    /* 0x30-0x3f */
        -1, 0, 1, 2,  3, 4, 5, 6,  7, 8, 9,10, 11,12,13,14,    /* 0x40-0x4f */
        15,16,17,18, 19,20,21,22, 23,24,25,-1, -1,-1,-1,63,    /* 0x50-0x5f */
        -1,26,27,28, 29,30,31,32, 33,34,35,36, 37,38,39,40,    /* 0x60-0x6f */
        41,42,43,44, 45,46,47,48, 49,50,51,-1, -1,-1,-1,-1,    /* 0x70-0x7f */
        -1,-1,-1,-1, -1,-1,-1,-1, -1,-1,-1,-1, -1,-1,-1,-1,    /* 0x80-0x8f */
        -1,-1,-1,-1, -1,-1,-1,-1, -1,-1,-1,-1, -1,-1,-1,-1,    /* 0x90-0x9f */
        -1,-1,-1,-1, -1,-1,-1,-1, -1,-1,-1,-1, -1,-1,-1,-1,    /* 0xa0-0xaf */
        -1,-1,-1,-1, -1,-1,-1,-1, -1,-1,-1,-1, -1,-1,-1,-1,    /* 0xb0-0xbf */
        -1,-1,-1,-1, -1,-1,-1,-1, -1,-1,-1,-1, -1,-1,-1,-1,    /* 0xc0-0xcf */
        -1,-1,-1,-1, -1,-1,-1,-1, -1,-1,-1,-1, -1,-1,-1,-1,    /* 0xd0-0xdf */
        -1,-1,-1,-1, -1,-1,-1,-1, -1,-1,-1,-1, -1,-1,-1,-1,    /* 0xe0-0xef */
        -1,-1,-1,-1, -1,-1,-1,-1, -1,-1,-1,-1, -1,-1,-1,-1,    /* 0xf0-0xff */
    };
    size_t max = b64url_get_decoded_buffer_size( len );

    if( (0 == max) || !enc || !out ) {
        return 0;
    }

    return decode( map, enc, len, out );
}


uint8_t* b64_decode_with_alloc( const uint8_t *enc, size_t len, size_t *out_len )
{
    return decode_w_alloc( b64_get_decoded_buffer_size, b64_decode, enc, len, out_len );
}


char* b64_encode_with_alloc( const uint8_t *raw, size_t len, size_t *out_len )
{
    return encode_w_alloc( b64_get_encoded_buffer_size, b64_encode, raw, len, out_len );
}


uint8_t* b64url_decode_with_alloc( const uint8_t *enc, size_t len, size_t *out_len )
{
    return decode_w_alloc( b64url_get_decoded_buffer_size, b64url_decode, enc, len, out_len );
}


char* b64url_encode_with_alloc( const uint8_t *raw, size_t len, size_t *out_len )
{
    return encode_w_alloc( b64url_get_encoded_buffer_size, b64url_encode, raw, len, out_len );
}


/*----------------------------------------------------------------------------*/
/*                             Internal functions                             */
/*----------------------------------------------------------------------------*/
static void encode( const char *map, const uint8_t *in, size_t len, uint8_t *out )
{
    uint32_t bits = 0;
    int bit_count = 0;
    size_t j = 0;

    for( size_t i = 0; i < len; i++ ) {
        bits = (bits << 8) | in[i];
        bit_count += 8;

        while( 6 <= bit_count ) {
            bit_count -= 6;
            out[j++] = (uint8_t) map[ 0x3f & (bits >> bit_count) ];
        }
    }

    /* Handle the extra bits. */
    if( bit_count ) {
        bits <<= 8;
        bit_count += 8;
        bit_count -= 6;
        out[j++] = (uint8_t) map[ 0x3f & (bits >> bit_count) ];
    }

    /* Pad */
    while( ('\0' != map[64]) && (0x03 & j) ) {
        out[j++] = (uint8_t) map[64];
    }
}


static size_t decode( const int8_t *map, const uint8_t *in, size_t len, uint8_t *out)
{

    uint32_t bits = 0;
    int bit_count = 0;
    size_t padding = 0;
    size_t j = 0;

    if( '=' == in[len - 1] ) {
        padding++;
        if( '=' == in[len - 2] ) {
            padding++;
        }

        /* If there is padding then it should only pad to ensure the string
         * has a multiple of 4.  Anything else is an error. */
        if( 0 != (0x03 & len) ) {
            return 0;
        }
    }

    len -= padding;

    for( size_t i = 0; i < len; i++ ) {
        int8_t val;

        val = map[in[i]];
        if( val < 0 ) {
            return 0;
        }
        bits = (bits << 6) | val;
        bit_count += 6;

        if( 8 <= bit_count ) {
            out[j++] = (uint8_t) (0x0ff & (bits >> (bit_count - 8)));
            bit_count -= 8;
        }
    }

    return j;
}

static uint8_t* decode_w_alloc( size_t (size_fn)(const size_t),
                                size_t (decode_fn)(const uint8_t*, const size_t, uint8_t*),
                                const uint8_t *enc, size_t len, size_t *out_len )
{
    size_t raw_len = size_fn( len );
    uint8_t *buf = NULL;

    if( out_len ) {
        *out_len = 0;
    }
    if( !raw_len || !enc || !out_len ) {
        return NULL;
    }

    buf = malloc( raw_len * sizeof(uint8_t) );
    if( buf ) {
        *out_len = decode_fn( enc, len, buf );
        if( 0 == *out_len ) {
            free( buf );
            buf = NULL;
        }
    }

    return buf;
}

static char* encode_w_alloc( size_t (size_fn)(const size_t),
                             void (encode_fn)(const uint8_t*, const size_t, uint8_t*),
                             const uint8_t *raw, size_t len, size_t *out_len )
{
    size_t enc_len = size_fn( len );
    char *buf = NULL;

    if( out_len ) {
        *out_len = 0;
    }

    if( !enc_len || !raw ) {
        return NULL;
    }

    buf = malloc( (enc_len + 1) * sizeof(char) );
    if( buf ) {
        encode_fn( raw, len, (uint8_t*) buf );
        buf[enc_len] = '\0';
        if( out_len ) {
            *out_len = enc_len;
        }
    }

    return buf;
}
